### ISSUE TITLE
Efficient Way to Pre-Download ERA5 inputs and Load them for Fine-Tuning

### ISSUE THREAD

**Issue by SasankhM** [microsoft/aurora#150]
I’m building on top of Aurora and adding custom decoders to predict additional variables. To do this, I still need to feed ERA5 data into Aurora to generate the latent representation before passing it to the custom decoders.

The [documentation](https://microsoft.github.io/aurora/example_era5.html) shows how to download ERA5 data for a specific day. However, what’s the best practice for pre-downloading all required ERA5 data so it can be efficiently loaded during model training, without becoming a bottleneck?

**Comment by wesselb** [comment]
Hey @SasankhM! It's very cool to hear that you're building on top of Aurora to build additional decoders. :)

To make training efficient, it is indeed best to download all training data to a local or cloud storage location and then load the data from there. This will be much much faster than downloading from the CDS on the fly. We unfortunately have not released any code to do this. Building a performant data loading pipeline is challening, but certainly worthwhile! I'm happy to give feedback/suggestions here should you run into any specific issues.

**Comment by SasankhM** [comment]
Awesome, thanks @wesselb! I think the community would definitely benefit from code to help with that at some point (maybe a future request).

In the meantime, do you have any guidance on how much storage space you’d recommend for keeping all the required ERA5 data? I’d like to size an instance appropriately or see if I can run it locally.

**Comment by SasankhM** [comment]
@wesselb, as a follow-up, I am noticing that loading in ERA5 data to then run Aurora inference can be slow/a bottleneck (I am using the loading process [described in the documentation here ](https://microsoft.github.io/aurora/example_era5.html), but it seems like the loading is not very efficient.  Have there been any optimizations on loading in ERA5 data whether for Aurora or even for a simple baseline that might not use Aurora? 

Also, curious if there is any time expectation for how much time inference should take on say an H100 for one Aurora batch (including loading ERA5 data)?

Thanks!!!

**Comment by wesselb** [comment]
@SasankhM The notebook is a simple example that's only meant to show some predictions by the model. You're right that it can't directly be scaled to a larger training run, because (amongst many other issues) data loading is too slow.

A simple way to optimise dataloading is the following:

1. Download your whole training dataset (e.g. ERA5) to a storage location that allows high download speeds (e.g. Azure Blob Storage). Important: Make sure that data for different timesteps are in different files!

2. Write a PyTorch `Dataset` that loads from the storage location.

3. Finally use PyTorch's `DataLoader` with `num_workers=2` or `num_workers=3` to do data loading asynchronously in your training loop.

See [these docs](https://docs.pytorch.org/tutorials/beginner/basics/data_tutorial.html) for some more info on PyTorch's `Dataset` and `DataLoader`.

> Also, curious if there is any time expectation for how much time inference should take on say an H100 for one Aurora batch (including loading ERA5 data)?

I believe that running the model forward and computing gradients should take around 3 seconds on an H100. You need approx. 600 MB per gradient step (context and target data of 300 MB). Assuming a download speed of around 10 Gbps, that means a download time of about 0.5 seconds (or 1 seconds for 5 Gbps download speed), which is well within the 3 seconds required for the forward and backward pass. The training loop should therefore definitely not be limited by data loading.

**Comment by SasankhM** [comment]
Thanks @wesselb! I do have 1. all ERA5 data downloaded on disk in the same place as where training is happening 2. have a PyTorch `Dataset` and `DataLoader` with workers for asynchronous data loading. 

Mainly interested in running just inference with Aurora to then use for custom decoders for my use case (so backward pass doesn't matter), so expecting about ~1 second per inference on an H100 feels reasonable. Still running into fairly slow training times (despite the decoders being fairly simple) on my end, so will investigate further! Thanks so much!

**Comment by wesselb** [comment]
Good luck!
